ports/nrf/common-hal/neopixel_write/__init__.c

```diff
@@ -0,0 +1,268 @@
+//   * Frequency: 800Khz
+// For 400Khz we just double the calculated times.
+// ---------- BEGIN Constants for the EasyDMA implementation -----------
+// The PWM starts the duty cycle in LOW. To start with HIGH we
+// need to set the 15th bit on each register.
+
+// WS2812 (rev A) timing is 0.35 and 0.7us
+//#define MAGIC_T0H               5UL | (0x8000) // 0.3125us
+//#define MAGIC_T1H              12UL | (0x8000) // 0.75us
+
+// WS2812B (rev B) timing is 0.4 and 0.8 us
+#define MAGIC_T0H               6UL | (0x8000) // 0.375us
+#define MAGIC_T1H              13UL | (0x8000) // 0.8125us
+#define CTOPVAL                20UL            // 1.25us
+
+// ---------- END Constants for the EasyDMA implementation -------------
+//
+// If there is no device available an alternative cycle-counter
+// implementation is tried.
+// The nRF52832 runs with a fixed clock of 64Mhz. The alternative
+// implementation is the same as the one used for the Teensy 3.0/1/2 but
+// with the Nordic SDK HAL & registers syntax.
+// The number of cycles was hand picked and is guaranteed to be 100%
+// organic to preserve freshness and high accuracy.
+// ---------- BEGIN Constants for cycle counter implementation ---------
+#define CYCLES_800_T0H  18  // ~0.36 uS
+#define CYCLES_800_T1H  41  // ~0.76 uS
+#define CYCLES_800      71  // ~1.25 uS
+
+// ---------- END of Constants for cycle counter implementation --------
+
+// find a free PWM device, which is not enabled and has no connected pins
+static NRF_PWM_Type* find_free_pwm (void) {
+    NRF_PWM_Type* PWM[3] = { NRF_PWM0, NRF_PWM1, NRF_PWM2 };
+
+    for ( int device = 0; device < 3; device++ ) {
+        if ( (PWM[device]->ENABLE == 0) && (PWM[device]->PSEL.OUT[0] & PWM_PSEL_OUT_CONNECT_Msk)
+                && (PWM[device]->PSEL.OUT[1] & PWM_PSEL_OUT_CONNECT_Msk)
+                && (PWM[device]->PSEL.OUT[2] & PWM_PSEL_OUT_CONNECT_Msk)
+                && (PWM[device]->PSEL.OUT[3] & PWM_PSEL_OUT_CONNECT_Msk) ) {
+            return PWM[device];
+        }
+    }
+
+    return NULL;
+}
+
+void common_hal_neopixel_write (const digitalio_digitalinout_obj_t* digitalinout, uint8_t *pixels, uint32_t numBytes) {
+    // To support both the SoftDevice + Neopixels we use the EasyDMA
+    // feature from the NRF25. However this technique implies to
+    // generate a pattern and store it on the memory. The actual
+    // memory used in bytes corresponds to the following formula:
+    //              totalMem = numBytes*8*2+(2*2)
+    // The two additional bytes at the end are needed to reset the
+    // sequence.
+    //
+    // If there is not enough memory, we will fall back to cycle counter
+    // using DWT
+    uint32_t pattern_size = numBytes * 8 * sizeof(uint16_t) + 2 * sizeof(uint16_t);
+    uint16_t* pixels_pattern = NULL;
+
+    NRF_PWM_Type* pwm = find_free_pwm();
+
+    // only malloc if there is PWM device available
+    if ( pwm != NULL ) {
+        pixels_pattern = (uint16_t *) m_malloc(pattern_size, false);
```

Comment:
Why malloc when it's freed below. Can't it be placed on the stack instead?

Reply:
since the numBytes is dynamic -> pattern_size is dynamic as well.

Reply:
You can allocate a variable-length array on the stack in C99: https://gcc.gnu.org/onlinedocs/gcc/Variable-Length.html

Reply:
I thought it is still used malloc under the hook ? btw, can the stack handle 12K bytes for 250 pixel ?

Reply:
I believe it does allocate it on the stack, But, 12kB on the stack, good point, maybe not. Advantage of using the stack is no fragmentation.

Reply:
@dhalbert  ah thanks, I know variable array declaration. But always thought compiler will use some kind of malloc under hook. Allocating in stack is new to me :D . Is it  compiler defined behavior or std behavior ?

Reply:
It's in the C99 spec for the C language. See that link above.

Reply:
Just read the text, I didn't see it say where the memory is located but it indeed implies the stack. Sorry for not being clear, I mean placing memory on stack is C99 specs or it allows compiler implement variable length array anyway it likes. Just Out of curiosity, since I don't use var len array much.

Reply:
never mind !! I guess knowing gcc behavior is good enough.It cover 95% of cases :D

Reply:
regarding malloc, will the alignment be multiple 4, or it can be odd. This example need the allocated mem is align of 2 for uint16_t

Reply:
Actually this calls the gc alloc, and a quick glance doesn't show me the alignment. But it's at least *4 aligned.

Reply:
Thanks, at least 4 is perfect, it will work with this.

Reply:
The gc allocates 16 byte blocks that are 4 byte aligned.